File: share_call_slots/call_slots.swift
import AppKit
import EventKit
import Foundation

let ist = TimeZone(identifier: "Asia/Kolkata")!
let minSlot: TimeInterval = 30 * 60
let buffer: TimeInterval = 15 * 60
let windowLength: TimeInterval = 12 * 60 * 60

func parseDays() -> Int {
    var days = 3
    var idx = 1
    let args = CommandLine.arguments
    while idx < args.count {
        let arg = args[idx]
        if arg == "--days", idx + 1 < args.count, let value = Int(args[idx + 1]), value > 0 {
            days = min(value, 14)
            idx += 2
            continue
        }
        if arg.hasPrefix("--days="), let value = Int(arg.dropFirst(7)), value > 0 {
            days = min(value, 14)
        }
        idx += 1
    }
    return days
}

func startOfISTDay(_ date: Date) -> Date {
    var calendar = Calendar(identifier: .gregorian)
    calendar.timeZone = ist
    return calendar.startOfDay(for: date)
}

func window(forISTEvening day: Date) -> (Date, Date) {
    let start = startOfISTDay(day).addingTimeInterval(17 * 60 * 60)
    return (start, start.addingTimeInterval(windowLength))
}

func mergeBusy(windowStart: Date, windowEnd: Date, busy: [(Date, Date)]) -> [(Date, Date)] {
    var clamped: [(Date, Date)] = []
    for (rawStart, rawEnd) in busy {
        let start = max(rawStart, windowStart)
        let end = min(rawEnd.addingTimeInterval(buffer), windowEnd)
        if end > start {
            clamped.append((start, end))
        }
    }
    clamped.sort { $0.0 < $1.0 }
    var merged: [(Date, Date)] = []
    for span in clamped {
        if let last = merged.last, span.0 <= last.1 {
            merged.removeLast()
            merged.append((last.0, max(last.1, span.1)))
        } else {
            merged.append(span)
        }
    }
    return merged
}

func freeSlots(windowStart: Date, windowEnd: Date, busy: [(Date, Date)]) -> [(Date, Date)] {
    var cursor = windowStart
    var free: [(Date, Date)] = []
    for (start, end) in mergeBusy(windowStart: windowStart, windowEnd: windowEnd, busy: busy) {
        if start.timeIntervalSince(cursor) >= minSlot {
            free.append((cursor, start))
        }
        cursor = max(cursor, end)
    }
    if windowEnd.timeIntervalSince(cursor) >= minSlot {
        free.append((cursor, windowEnd))
    }
    return free
}

func formatIST(_ date: Date, template: String) -> String {
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = ist
    formatter.dateFormat = template
    return formatter.string(from: date)
}

func buildMessage(days: [(Date, [(Date, Date)])]) -> String {
    var lines = [
        "Available for calls",
        "Calendar: saudagercash14@gmail.com",
        "Window: 5:00 PM – 5:00 AM IST",
        "(6:30 AM – 6:30 PM US Central)",
        "",
        "Pick a slot and reply here.",
        "",
    ]
    var any = false
    for (day, slots) in days {
        guard !slots.isEmpty else { continue }
        any = true
        lines.append("*\(formatIST(day, template: "EEE, d MMM"))*")
        for (start, end) in slots {
            lines.append(
                "• \(formatIST(start, template: "h:mm a")) – \(formatIST(end, template: "h:mm a")) IST"
            )
        }
        lines.append("")
    }
    if !any {
        lines.append("No free slots in this window.")
    }
    return lines.joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines) + "\n"
}

func requestAccess(_ store: EKEventStore) -> Bool {
    let semaphore = DispatchSemaphore(value: 0)
    var granted = false
    if #available(macOS 14.0, *) {
        store.requestFullAccessToEvents { ok, _ in
            granted = ok
            semaphore.signal()
        }
    } else {
        store.requestAccess(to: .event) { ok, _ in
            granted = ok
            semaphore.signal()
        }
    }
    _ = semaphore.wait(timeout: .now() + 30)
    return granted
}

func loadBusy(store: EKEventStore, start: Date, end: Date) -> [(Date, Date)] {
    let calendars = store.calendars(for: .event)
    let predicate = store.predicateForEvents(withStart: start, end: end, calendars: calendars)
    return store.events(matching: predicate).compactMap { event in
        if event.status == .canceled { return nil }
        if event.availability == .free { return nil }
        return (event.startDate, event.endDate)
    }
}

let dayCount = parseDays()
let store = EKEventStore()
guard requestAccess(store) else {
    fputs("Calendar access was denied. Enable Calendar for Terminal/Shortcuts in System Settings → Privacy & Security → Calendars.\n", stderr)
    exit(1)
}

var calendar = Calendar(identifier: .gregorian)
calendar.timeZone = ist
let today = startOfISTDay(Date())
let rangeStart = today.addingTimeInterval(-12 * 60 * 60)
let rangeEnd = today.addingTimeInterval(TimeInterval((dayCount + 2) * 24 * 60 * 60))
let busy = loadBusy(store: store, start: rangeStart, end: rangeEnd)

var perDay: [(Date, [(Date, Date)])] = []
for offset in 0..<dayCount {
    guard let day = calendar.date(byAdding: .day, value: offset, to: today) else { continue }
    let (windowStart, windowEnd) = window(forISTEvening: day)
    perDay.append((day, freeSlots(windowStart: windowStart, windowEnd: windowEnd, busy: busy)))
}

let message = buildMessage(days: perDay)
print(message, terminator: "")
NSPasteboard.general.clearContents()
NSPasteboard.general.setString(message, forType: .string)

if CommandLine.arguments.contains("--whatsapp") {
    var allowed = CharacterSet.urlQueryAllowed
    allowed.remove(charactersIn: ":/?#[]@!$&'()*+,;=")
    let encoded = message.addingPercentEncoding(withAllowedCharacters: allowed) ?? ""
    if let url = URL(string: "whatsapp://send?text=\(encoded)") {
        NSWorkspace.shared.open(url)
    }
}

File: share_call_slots/call_slots.py
#!/usr/bin/env python3
"""Read Apple Calendar, print IST call slots, copy them, optionally open WhatsApp."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import urllib.parse
from datetime import datetime, timedelta
from pathlib import Path
from zoneinfo import ZoneInfo

from slots import IST, evening_date_for, format_whatsapp, free_slots, window_for_ist_evening

JXA = r"""
ObjC.import("EventKit")
ObjC.import("Foundation")

function iso(date) {
  const fmt = $.NSISO8601DateFormatter.alloc.init
  fmt.formatOptions = $.NSISO8601DateFormatWithInternetDateTime + $.NSISO8601DateFormatWithFractionalSeconds
  return ObjC.unwrap(fmt.stringFromDate(date))
}

function run(argv) {
  const startISO = argv[0]
  const endISO = argv[1]
  const store = $.EKEventStore.alloc.init
  const allCalendars = store.calendarsForEntityType($.EKEntityTypeEvent)
  const wanted = "saudagercash14@gmail.com"
  const calendars = $.NSMutableArray.alloc.init
  for (let i = 0; i < allCalendars.count; i++) {
    const cal = allCalendars.objectAtIndex(i)
    const title = ObjC.unwrap(cal.title) || ""
    if (title === wanted || title.indexOf(wanted) >= 0) {
      calendars.addObject(cal)
    }
  }
  if (calendars.count === 0) {
    throw new Error("Calendar saudagercash14@gmail.com was not found")
  }
  const fmt = $.NSISO8601DateFormatter.alloc.init
  fmt.formatOptions = $.NSISO8601DateFormatWithInternetDateTime
  const start = fmt.dateFromString(startISO)
  const end = fmt.dateFromString(endISO)
  if (start === null || end === null) {
    throw new Error("Could not parse date range")
  }
  const predicate = store.predicateForEventsWithStartDateEndDateCalendars(start, end, calendars)
  const events = store.eventsMatchingPredicate(predicate)
  const lines = []
  for (let i = 0; i < events.count; i++) {
    const event = events.objectAtIndex(i)
    if (event.status === 2) continue
    if (event.availability === 1) continue
    lines.push(iso(event.startDate) + "\t" + iso(event.endDate))
  }
  return lines.join("\n")
}
"""


def parse_iso(value: str) -> datetime:
    text = value.strip()
    if text.endswith("Z"):
        text = text[:-1] + "+00:00"
    return datetime.fromisoformat(text).astimezone(IST)


def load_busy(start: datetime, end: datetime) -> list[tuple[datetime, datetime]]:
    auth = subprocess.run(
        [
            "osascript",
            "-l",
            "JavaScript",
            "-e",
            """
ObjC.import("EventKit")
const store = $.EKEventStore.alloc.init
const status = $.EKEventStore.authorizationStatusForEntityType($.EKEntityTypeEvent)
const calendars = store.calendarsForEntityType($.EKEntityTypeEvent)
JSON.stringify({status: Number(status), calendars: Number(calendars.count)})
""",
        ],
        capture_output=True,
        text=True,
        check=False,
    )
    try:
        info = json.loads(auth.stdout.strip().splitlines()[-1])
    except Exception:
        info = {"status": -1, "calendars": 0}
    if int(info.get("calendars") or 0) <= 0:
        raise RuntimeError(
            "This Python helper cannot read Apple Calendar from Cursor/Terminal yet "
            "(no EventKit permission).\n\n"
            "Use the installed Shortcut instead:\n"
            "  Mac: open Shortcuts and run “Share Call Slots”\n"
            "  iPhone: wait for iCloud sync, or AirDrop\n"
            f"    {Path(__file__).resolve().parent / 'out' / 'Share Call Slots.shortcut'}"
        )
    result = subprocess.run(
        [
            "osascript",
            "-l",
            "JavaScript",
            "-e",
            JXA,
            "--",
            start.astimezone(ZoneInfo("UTC")).strftime("%Y-%m-%dT%H:%M:%SZ"),
            end.astimezone(ZoneInfo("UTC")).strftime("%Y-%m-%dT%H:%M:%SZ"),
        ],
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        err = (result.stderr or result.stdout).strip()
        raise RuntimeError(
            "Could not read Apple Calendar. Allow Calendar access for Terminal "
            f"in System Settings → Privacy & Security → Calendars.\n{err}"
        )
    busy: list[tuple[datetime, datetime]] = []
    for line in result.stdout.splitlines():
        if "\t" not in line:
            continue
        raw_start, raw_end = line.split("\t", 1)
        busy.append((parse_iso(raw_start), parse_iso(raw_end)))
    return busy


def build_message(days: int = 1) -> str:
    now = datetime.now(IST)
    day = evening_date_for(now)
    window_start, window_end = window_for_ist_evening(day)
    busy = load_busy(window_start - timedelta(hours=12), window_end + timedelta(hours=12))
    slots = free_slots(window_start, window_end, busy)
    return format_whatsapp([(day, slots)], window_start=window_start, window_end=window_end)


def copy_to_clipboard(text: str) -> None:
    subprocess.run(["pbcopy"], input=text.encode("utf-8"), check=True)


def open_whatsapp(text: str) -> None:
    encoded = urllib.parse.quote(text, safe="")
    subprocess.run(["open", f"whatsapp://send?text={encoded}"], check=False)


def main() -> int:
    parser = argparse.ArgumentParser(description="Share free call slots from Apple Calendar")
    parser.add_argument("--days", type=int, default=1)
    parser.add_argument("--whatsapp", action="store_true")
    args = parser.parse_args()
    try:
        message = build_message(max(1, min(args.days, 14)))
    except RuntimeError as exc:
        print(exc, file=sys.stderr)
        return 1
    sys.stdout.write(message)
    copy_to_clipboard(message)
    if args.whatsapp:
        open_whatsapp(message)
    return 0


if __name__ == "__main__":
    sys.path.insert(0, str(Path(__file__).resolve().parent))
    raise SystemExit(main())
